### ISSUE TITLE
Decimal type is missing precision and scale

### ISSUE THREAD

**Issue by ClassyCircuit** [ReversePOCO/EntityFramework-Reverse-POCO-Code-First-Generator#411]
Hi,
For decimal type there are no properties generated, which would indicate how many numbers are allowed for this field.
In our database we have limits set for each decimal field:
![image](https://user-images.githubusercontent.com/6424248/39232794-20bbfa98-4877-11e8-911c-61bb47c4380e.png)
I would like to have a property, the same way you have made for nvarchar, that would tell what are the limits for each decimal field. 
Right now it generates the following:

```
[Column(@"PL_GrossAmount", Order = 7, TypeName = "decimal")]
[Required]
[Display(Name = "Pl gross amount")]
public decimal PlGrossAmount { get; set; } // PL_GrossAmount
```

Thanks

**Comment by agentKnipe** [comment]
@sjh37 is there a way to handle this for stored procedure return models? We have handled it with the tables returns using the  Settings.UpdateColumn delegate and set the necessary precision and scale, but we havent been able to find where to set this for stored procedure return models.

**Comment by sjh37** [comment]
Hi @janissimsons Is this for EF Core? EF6 already supports the `.HasPrecision(precision, scale)` however support for EF Core was lacking.
This has now been added to EF Core just 11 days ago, but so far has not yet been released. See [https://github.com/dotnet/efcore/pull/20468](https://github.com/dotnet/efcore/pull/20468)

**Comment by agentKnipe** [comment]
I was wondering if there was any movement on this for .net core? I saw a fix for .net five, and have pulled down the newest version of reverse poco (and updated the ttinclude to 3.13.0) but my model isnt reflecting the decimal columns still.  is there something needed in the database.tt to get this to work?

**Comment by sjh37** [comment]
Hi @agentKnipe 

**TL;DR**
It should be working fine. I've added some unit tests in a new file called `DecimalPrecisionTests` to prove it's working.
It could be your `<database>.tt` file is out of date with writing the data annotations corrections. To make this easier, I've created a helper function in settings: `Settings.ApplyDataAnnotations` which you can call it in your `Settings.UpdateColumn` delegate in your `<database>.tt` file so it can be simplified to:

```csharp
Settings.UpdateColumn = delegate(Column column, Table table, List<EnumDefinition> enumDefinitions, List<JsonColumnMapping> jsonColumnMappings)
{
    Settings.ApplyJsonPropertyNameAttribute(column);
    Settings.ApplyJsonColumnMappings(column, table, jsonColumnMappings);
    Settings.ApplyDataAnnotations(column);  // <<-----------------------------------
    Settings.ApplyEnumTypeReplacement(column, table, enumDefinitions);
}
```

**Current behaviour (EF Core 8, 9, 10)**

The generator already emits precision and scale for `decimal`/`numeric` columns. For a column defined as `decimal(18,4)` you will get:

*Fluent API mode:*
```csharp
builder.Property(x => x.PlGrossAmount)
    .HasColumnName(@"PL_GrossAmount")
    .HasColumnType("decimal(18,4)")
    .HasPrecision(18,4)
    .IsRequired();
```

*Data annotations mode:*
```csharp
[Precision(18, 4)]
public decimal PlGrossAmount { get; set; }
```
(with `.HasColumnType("decimal(18,4)")` still written into the fluent config file)

**EF6**

EF6 emits `.HasPrecision(18,4)` in the model builder configuration.

**Historical context**

This issue was originally filed in 2018 and labelled "Awaiting Microsoft EF Core update" because early EF Core had no `HasPrecision` API. That API arrived in EF Core 5; the generator was subsequently updated to use it. The `[Precision]` data annotation (EF Core 6+) is also generated where applicable.

**What was added in this fix**

Unit tests were missing that would have caught any regression here. A new `DecimalPrecisionTests` fixture has now been added, covering all combinations:

- EF Core fluent API - precision in both `HasColumnType` and `HasPrecision`
- EF Core data annotations - `[Precision]` attribute, no duplicate `HasPrecision` in config
- EF6 - `HasPrecision` in config
- Both `decimal` and `numeric` SQL types
- Zero precision/scale (no suffix appended)

If you are still seeing `TypeName = "decimal"` without precision, please share your generator version, template type, and database type, and I'll take a closer look.

**Comment by sjh37** [comment]
Closing this ticket. Let me know if you still have issues, and I can re-open it.

**Comment by agentKnipe** [comment]
ok, let me work through this a bit, Im definitely missing the column update from my database.tt.  Should I expect the same behavior for a sproc return model?

**Comment by sjh37** [comment]
**Re: sproc return model precision/scale**

Short answer: **No**, you cannot expect the same `[Precision]`/`HasPrecision` behavior for stored procedure return models - this is a fundamental limitation in how the generator discovers sproc return columns.

**Why table columns work**

For tables and views, the generator queries `information_schema.columns` / `sys.columns` directly and reads the SQL type definition including precision and scale. Those values are stored on the `Column` object and fed into `Settings.ApplyDataAnnotations`, `HasPrecision`, etc.

**Why sproc return models cannot**

For sprocs, the generator issues `SET FMTONLY ON` and calls `FillSchema()` to get an ADO.NET `DataSet`. This gives back `System.Data.DataColumn` objects. When SQL Server maps `decimal(18,4)` through ADO.NET, the `DataColumn` only records `DataType = typeof(System.Decimal)` - the precision and scale are discarded. There is no precision/scale property on `DataColumn`, so `WriteStoredProcReturnColumn` has nothing to work from and can only emit `decimal` or `decimal?`.

**What can be done**

The proper fix would be to use `sys.dm_exec_describe_first_result_set` for SQL Server instead of (or alongside) `SET FMTONLY ON` - that DMV returns `precision` and `scale` for each result column, which would let the generator emit the same quality annotations as it does for tables. I'll track this as a separate enhancement.

In the meantime, there is no built-in hook equivalent to `Settings.UpdateColumn` for sproc return model columns, so there is no workaround short of editing the generated return model class by hand after generation.

For `sp_describe_first_result_set`  you can track issue #448 which I'll bump up my todo list 🥇

**Comment by agentKnipe** [comment]
sounds great, thank you for the in depth explanation.  ill keep an eye out for the enhancement.

**Comment by sjh37** [comment]
Released in v3.14.0
